zeppelin-zengine/src/main/java/org/apache/zeppelin/interpreter/InterpreterSetting.java

```diff
@@ -713,16 +719,101 @@ RemoteInterpreterProcess createInterpreterProcess() {
     return remoteInterpreterProcess;
   }
 
+  private boolean isSparkConf(String key, String value) {
+    return !StringUtils.isEmpty(key) && key.startsWith("spark.") && !StringUtils.isEmpty(value);
+  }
+
   private Map<String, String> getEnvFromInterpreterProperty(Properties property) {
-    Map<String, String> env = new HashMap<>();
-    for (Object key : property.keySet()) {
-      if (RemoteInterpreterUtils.isEnvString((String) key)) {
-        env.put((String) key, property.getProperty((String) key));
+    Map<String, String> env = new HashMap<String, String>();
+    Properties sparkProperties = new Properties();
+    String sparkMaster = "local[*]";
+    for (String key : property.stringPropertyNames()) {
+      if (RemoteInterpreterUtils.isEnvString(key)) {
+        env.put(key, property.getProperty(key));
       }
+      if (key.equals("master")) {
+        sparkMaster = property.getProperty("master");
+      }
+      if (isSparkConf(key, property.getProperty(key))) {
+        sparkProperties.setProperty(key, toShellFormat(property.getProperty(key)));
+        if (key.equals("spark.master")) {
+          if (sparkMaster == null) {
+            sparkMaster = property.getProperty(key);
+          }
+        }
+      }
+    }
+
+    setupPropertiesForPySpark(sparkProperties, sparkMaster);
+    setupPropertiesForSparkR(sparkProperties, sparkMaster, property.getProperty("SPARK_HOME"));
+    if (sparkMaster != null && sparkMaster.equals("yarn-cluster")) {
+      env.put("SPARK_YARN_CLUSTER", "true");
     }
+
+    StringBuilder sparkConfBuilder = new StringBuilder();
+    if (sparkMaster != null) {
+      sparkConfBuilder.append(" --master " + sparkMaster);
+    }
+    if (sparkMaster.equals("yarn-cluster")) {
```

Comment:
ditto

Reply:
No, this only works for yarn-cluster mode, not yarn-client mode. yarn-client mode still use the default logj4.properties, because driver runs in the zeppelin host

Reply:
well, my point is we are not handling yarn cluster mode consistently.
yarn cluster mode can set by master=yarn, deployMode=cluster or
master=yarn-cluster (deprecated)
we have both in this PR.. either we should support both, or the supported not deprecated way, I think

Reply:
Right, I was planning to fix it in the next PR. This is a messy. Let me fix it in this PR.